# Introduction: the Titanic Survivals dataset

As seen in the previous practical lesson, Pandas is a library for data manipulation and analysis. In this lesson we'll use the functionalities of Pandas in order to analyze and clean up the data of the Titanic Survivals dataset.

In [1]:
import pandas as pd
import numpy as np

df = pd.read_csv('http://ailab.uniud.it/wp-content/uploads/2020/10/titanic_survivals.csv', sep=',', index_col=None)  # Loading the dataset in a Pandas Dataframe
print(df.head(10))  # Printing the first ten values of the dataset

   PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund; Mr. Owen Harris    male  22.0      1   
1  Cumings; Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen; Miss. Laina  female  26.0      0   
3       Futrelle; Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen; Mr. William Henry    male  35.0      0   
5                                   Moran; Mr. James    male   NaN      0   
6                            McCarthy; Mr. Timothy J    male  54

By printing the name of the dataset columns we notice the dataset has twelve features, containing the following data: 

*   PassengerId: identifier of the passenger
*   Survived: whether the passenger survived (1) or not (0)
*   Pclass:	ticket class (1 = 1st, 2 = 2nd, 3 = 3rd class)
*   Sex: sex of the passenger
*   Age: age in years
*   Sibsp: number of siblings/spouses aboard the Titanic	
*   Parch: number of parents/children aboard the Titanic	
*   Ticket: ticket number
*   Fare: passenger fare
*   Cabin: cabin number
*   Embarked: port of Embarkation (C=Cherbourg, Q=Queenstown, S=Southampton)


In [2]:
print(df.columns)

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')


# Analyzing the features of the dataset

Pandas provides a series of commands to analyze the statistics related with the columns of a dataframe. For instance, with the command *values* the user can print the values of a selected dataframe column:

In [3]:
name_of_the_passengers = df["Name"].values
name_of_the_passengers = df["Survived"].values

print(name_of_the_passengers[:10]) # Printing the names of the first 10 passengers

[0 1 1 1 0 0 0 0 1 1]


Moreover, applying the commands *min()* and *max()* to the columns of a dataframe which contain numerical values allows us to see the minimal and maximal values of such columns. In case a column contains text data, *min()* and *max()* return the first and last values of the lexicographical order of the values in the column.

In [4]:
min_val_passengerid = df["PassengerId"].min()  # Calculating the minimal value of the feature "PassengerID"
print(f"Minimal value for column \"PassengerId\":{min_val_passengerid}")
max_val_passengerid = df["PassengerId"].max()  # Calculating the maximal value of the feature "PassengerID"
print(f"Maximal value for column \"PassengerId\":{max_val_passengerid}")

min_val_name = df["Name"].values.min()
print(f"Minimal value for column \"Name\":{min_val_name}")
max_val_name = df["Name"].values.max()
print(f"Maximal value for column \"Name\":{max_val_name}")

min_val_age = df["Age"].min()
print(f"Minimal value for column \"Age\":{min_val_age}")
max_val_age = df["Age"].max()
print(f"Maximal value for column \"Age\":{max_val_age}")

Minimal value for column "PassengerId":1
Maximal value for column "PassengerId":891
Minimal value for column "Name":Abbing; Mr. Anthony
Maximal value for column "Name":van Melkebeke; Mr. Philemon
Minimal value for column "Age":0.42
Maximal value for column "Age":80.0


With the command *value_counts()* the user can print both the values of a feature and the number of occurrences of such values in the feature.
Use *value_counts()* to print the values and the related number of occurrences of the feature "Survived":

In [5]:
values = df["Survived"].value_counts()

print(values)

0    549
1    342
Name: Survived, dtype: int64


With *value_counts()* we now know that, of the 891 passengers contained in the dataset, only 342 survived while 549 died.

Such statistics isn't very informative, so let's see how the number of survivors and of victims is distributed among men and women. In order to do so, we need to first select all the female and male passengers of the dataset and then use again the *value_counts()* function to count how many survived and how many didn't.

In [6]:
#value_counts() not works
survived_passengers = df[["Survived","Sex"]]  # Selecting only the columns we need from the dataset

survived_passengers_females = survived_passengers[survived_passengers["Sex"] == "female"]  # Selecting the female passengers
print(survived_passengers_females)
#print(survived_passengers_females.value_counts())  # Counting the survived and dead female passengers

survived_passengers_males = survived_passengers[survived_passengers["Sex"] == "male"]
print(survived_passengers_males)
#print(survived_passengers_males.value_counts())  # Counting the survived and dead male passengers

     Survived     Sex
1           1  female
2           1  female
3           1  female
8           1  female
9           1  female
..        ...     ...
880         1  female
882         0  female
885         0  female
887         1  female
888         0  female

[314 rows x 2 columns]
     Survived   Sex
0           0  male
4           0  male
5           0  male
6           0  male
7           0  male
..        ...   ...
883         0  male
884         0  male
886         0  male
889         1  male
890         0  male

[577 rows x 2 columns]


Now, let's visualize with *matplotlib* the data we just obtained in the last two examples.

In [7]:
import matplotlib.pyplot as plt

data_to_visualize = survived_passengers
labels = ["Survived", "Died"]

women_survived =  survived_passengers_females[survived_passengers_females["Survived"] == 1]# Selecting from survived_passengers_females only the female passengers that survived
number_of_women_survived = len(women_survived)
women_died =  survived_passengers_females[survived_passengers_females["Survived"] == 0]# Counting from survived_passengers_females how many female passengers died
number_of_women_died = len(women_died)

men_survived =  survived_passengers_males[survived_passengers_males["Survived"] == 1]# Selecting from survived_passengers_males only the male passengers that survived
number_of_men_survived = len(men_survived)
men_died = survived_passengers_males[survived_passengers_males["Survived"] == 0]# Selecting from survived_passengers_males only the male passengers that died
number_of_men_died = len(men_died)

x = np.arange(len(labels))  # the label locations
width = 0.4  # the width of the bars

fig, ax = plt.subplots()
rects1 = ax.bar(x - width/2, [number_of_men_survived, number_of_women_survived], width, label='Men')
rects2 = ax.bar(x + width/2, [number_of_men_died, number_of_women_died], width, label='Women')

# Add some text for labels, title and custom x-axis tick labels, etc.
ax.set_ylabel('Passengers')
ax.set_title('Survived and dead passengers by gender')
ax.set_xticks(x)
ax.set_xticklabels(labels)
ax.legend()

Now, let's see the median and mean age of both the men and women that survived and died through the *median()* and *mean()* commands. Let's first see the ages of all the male and female passengers of the Titanic.

In [8]:
female_passengers = df[df["Sex"] == "female"]  # Selecting only the female passengers
female_passengers_age = female_passengers["Age"]  # Selecting the age of the female passengers
print(female_passengers_age, "\n")

male_passengers =  df[df["Sex"] == "male"]# Selecting only the male passengers
male_passengers_age =  male_passengers["Age"]# Selecting the age of the male passengers
print(male_passengers_age)

1      38.0
2      26.0
3      35.0
8      27.0
9      14.0
       ... 
880    25.0
882    22.0
885    39.0
887    19.0
888     NaN
Name: Age, Length: 314, dtype: float64 

0      22.0
4      35.0
5       NaN
6      54.0
7       2.0
       ... 
883    28.0
884    25.0
886    27.0
889    26.0
890    32.0
Name: Age, Length: 577, dtype: float64


In [9]:
passengers_survived = df[df["Survived"] == 1]  # Selecting from the whole dataset only the survived passengers
passengers_died = df[df["Survived"] == 0]  # Selecting from the whole dataset only the passengers who died

female_survived_passengers = passengers_survived[passengers_survived["Sex"] == "female"] # Selecting only the female survived passengers
female_dead_passengers =  passengers_died[passengers_died["Sex"] == "female"]# Selecting only the female passengers that died

male_survived_passengers = passengers_survived[passengers_survived["Sex"] == "male"] # Selecting only the male survived passengers
male_dead_passengers =  passengers_died[passengers_died["Sex"] == "male"]# Selecting only the male passengers that died

female_survived_passengers_age = female_survived_passengers["Age"] # Selecting only the age of the female survived passengers
female_dead_passengers_age =  female_dead_passengers["Age"]# Selecting only the age of the female passengers that died

male_survived_passengers_age =  male_survived_passengers["Age"]# Selecting only the age of the female survived passengers
male_dead_passengers_age =  male_dead_passengers["Age"]# Selecting only the age of the male passengers that died

median_age_survived_female_passengers = np.mean(female_survived_passengers_age)# Calulating the median age of the survived female passengers
print(f"Median age of the survived female passengers:{median_age_survived_female_passengers}")
median_age_dead_female_passengers = np.mean(female_dead_passengers_age)# Calulating the median age of the female passengers that died
print(f"Median age of the female passengers that died:{median_age_dead_female_passengers}")

median_age_survived_male_passengers = np.mean(male_survived_passengers_age)  # Calulating the median age of the survived male passengers
print(f"Median age of the survived male passengers:{median_age_survived_male_passengers}")
median_age_dead_male_passengers =  np.mean(male_dead_passengers_age) # Calulating the median age of the male passengers that died
print(f"Median age of the male passengers that died:{median_age_dead_male_passengers}")

Median age of the survived female passengers:28.84771573604061
Median age of the female passengers that died:25.046875
Median age of the survived male passengers:27.276021505376345
Median age of the male passengers that died:31.618055555555557


Now, let's try analyzing the percentage of survivals of the Titanic passengers depending on the port from which they embarked. Firstly, let's group the passengers by the port they embarked from using the function *groupby()*.

In [18]:
groups_passengers_by_port = df.groupby("Embarked")

for key, item in groups_passengers_by_port:
  group = groups_passengers_by_port.get_group(key)  # Getting the group
  print(group)  # Printing each group by accessing them through their key

     PassengerId  Survived  Pclass  \
1              2         1       1   
9             10         1       2   
19            20         1       3   
26            27         0       3   
30            31         0       1   
..           ...       ...     ...   
866          867         1       2   
874          875         1       2   
875          876         1       3   
879          880         1       1   
889          890         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings; Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
9                  Nasser; Mrs. Nicholas (Adele Achem)  female  14.0      1   
19                             Masselmani; Mrs. Fatima  female   NaN      0   
26                             Emir; Mr. Farred Chehab    male   NaN      0   
30                            Uruchurtu; Don. Manuel E    male  40.0      0   
..                                                 ...     ...   ... 

Secondly, let's calculate how many passengers died and survived for each group, and save such values in a corresponding dictionary.

In [11]:
survived_by_port = {}
dead_by_port = {}

full_name_ports = {"C":"Cherbourg", "Q":"Queenstown", "S":"Southampton"}  # Dictionary mapping the values in the dataset with the full names of the ports

for key, item in groups_passengers_by_port:

  full_name_port =  full_name_ports[key]# Getting the full name of the port by accessing to the dictionary

  group =  groups_passengers_by_port.get_group(key)# Getting the group of passengers embarked from the currently selected port
  
  group_survived =  group[group["Survived"]==1]# Getting the subset of survived passengers by group
  survived_by_port[key] = len(group_survived)  # Getting the number of survived passengers by group

  group_died =  group[group["Survived"]==0]# Getting the subset of passengers that died by group
  dead_by_port[key] = len(group_died) # Getting the number of passengers that died by group

  print(f"Number of survived passengers embarked from {full_name_port}: {survived_by_port[key]}")
  print(f"Number of passengers that died and embarked from {full_name_port}: {dead_by_port[key]}")
print(dead_by_port)

Number of survived passengers embarked from Cherbourg: 93
Number of passengers that died and embarked from Cherbourg: 75
Number of survived passengers embarked from Queenstown: 30
Number of passengers that died and embarked from Queenstown: 47
Number of survived passengers embarked from Southampton: 217
Number of passengers that died and embarked from Southampton: 427
{'C': 75, 'Q': 47, 'S': 427}


Lastly, for each port we calculate the percentages of all the passengers that departed from there and then survived or died during the sinking of the Titanic.

In [12]:
for embarkment_port in survived_by_port:

  full_name_port = full_name_ports[embarkment_port]  # Getting the full name of the port, dict

  number_of_survived = survived_by_port[embarkment_port] # Getting the number of survived passengers
  number_of_dead =  dead_by_port[embarkment_port]# Getting the number of dead passengers
 
  total_of_passengers_from_port = len(groups_passengers_by_port.get_group(embarkment_port))  # Getting the total of passengers departed from a port
  percentage_of_survived = number_of_survived/total_of_passengers_from_port  # Getting the percentage of passengers departed from the same embarkment port that survived
  percentage_of_dead = number_of_dead/total_of_passengers_from_port  # Getting the percentage of passengers departed from the same embarkment port that died

  percentage_of_survived = round(percentage_of_survived, 2)  # Rounding the percentage
  percentage_of_dead = round(percentage_of_dead, 2)

  print(f"Percentage of passengers that departed from {full_name_port} and survived: {100*percentage_of_survived}%")
  print(f"Percentage of passengers that departed from {full_name_port} and died: {100*percentage_of_dead}%\n")

Percentage of passengers that departed from Cherbourg and survived: 55.00000000000001%
Percentage of passengers that departed from Cherbourg and died: 45.0%

Percentage of passengers that departed from Queenstown and survived: 39.0%
Percentage of passengers that departed from Queenstown and died: 61.0%

Percentage of passengers that departed from Southampton and survived: 34.0%
Percentage of passengers that departed from Southampton and died: 66.0%



Through this calculations we now know that around the 55% of the passengers that departed from Cherbourg survived, while only the 39% and 34% of passengers that departed from respectively Queenstown and Southamptom survived.

But what if we want to enstablish whether or not there's a hidden association between the features of a dataset? To do so there's the the *corr* function of Pandas, which calculates the correlation between each pair of columns of the dataset. The default method use by *corr* to calculate the correlation is the Pearson's standard correlation coefficient, which has a value **between -1 and 1**.
If the coefficient between two columns is 1 the two columns have a **perfect positive linear correlation**, while if it's -1 they have a **perfect negative linear correlation**; if the coefficient is 0 the two columns have **no correlation**.

In [13]:
correlations = df.corr() # Calculating the correlations between the columns of the dataset 

print(correlations)

             PassengerId  Survived    Pclass       Age     SibSp     Parch  \
PassengerId     1.000000 -0.005007 -0.035144  0.036847 -0.057527 -0.001652   
Survived       -0.005007  1.000000 -0.338481 -0.077221 -0.035322  0.081629   
Pclass         -0.035144 -0.338481  1.000000 -0.369226  0.083081  0.018443   
Age             0.036847 -0.077221 -0.369226  1.000000 -0.308247 -0.189119   
SibSp          -0.057527 -0.035322  0.083081 -0.308247  1.000000  0.414838   
Parch          -0.001652  0.081629  0.018443 -0.189119  0.414838  1.000000   
Fare            0.012658  0.257307 -0.549500  0.096067  0.159651  0.216225   

                 Fare  
PassengerId  0.012658  
Survived     0.257307  
Pclass      -0.549500  
Age          0.096067  
SibSp        0.159651  
Parch        0.216225  
Fare         1.000000  


Typically when working with datasets the correlation is used in order to find out whether two features of the dataset have redundant data. If a pair of features have a very high correlation value (≥0.9 or ≤-0.9) one of the two features is dropped, because it doesn't add any information to that already contained in the other column. Now, let's find out if there are pairs of features in our dataset with a very high correlation with the *where* function of Pandas:

In [14]:
for column in correlations.columns:
  print(f"Correlations in column {column}:\n{correlations[column]}\n")

Correlations in column PassengerId:
PassengerId    1.000000
Survived      -0.005007
Pclass        -0.035144
Age            0.036847
SibSp         -0.057527
Parch         -0.001652
Fare           0.012658
Name: PassengerId, dtype: float64

Correlations in column Survived:
PassengerId   -0.005007
Survived       1.000000
Pclass        -0.338481
Age           -0.077221
SibSp         -0.035322
Parch          0.081629
Fare           0.257307
Name: Survived, dtype: float64

Correlations in column Pclass:
PassengerId   -0.035144
Survived      -0.338481
Pclass         1.000000
Age           -0.369226
SibSp          0.083081
Parch          0.018443
Fare          -0.549500
Name: Pclass, dtype: float64

Correlations in column Age:
PassengerId    0.036847
Survived      -0.077221
Pclass        -0.369226
Age            1.000000
SibSp         -0.308247
Parch         -0.189119
Fare           0.096067
Name: Age, dtype: float64

Correlations in column SibSp:
PassengerId   -0.057527
Survived      -0.03532

As we can see, there are no pairs of columns with a particularly high correlation index.